### ISSUE TITLE
BMW 520d F11 2016

### ISSUE THREAD

**Issue by iz3man** [m1n3rva/bimmer_connected#69]
No data shown
I'm on my phone, so can't attach files. Formating is difficult too. Hope all is readable.

**Comment by iz3man** [comment]
status_0.txt

{
  "vehicleStatus": {
    "DCS_CCH_Activation": "NA",
    "DCS_CCH_Ongoing": false,
    "position": {
      "lat": 12.3456,
      "lon": 34.5678,
      "status": "OK"
    },
    "steering": "LH",
    "updateTime": "2018-03-06T12:21:34+0100",
    "vin": "some_vin"
  }

vehicles_0.txt

{
  "vehicles": [
    {
      "a4a": "USB_ONLY",
      "bodytype": "F11",
      "brand": "BMW",
      "breakdownNumber": "+4989358957103",
      "carCloud": "NOT_SUPPORTED",
      "chargeNow": "NOT_SUPPORTED",
      "chargingControl": "NOT_SUPPORTED",
      "climateControl": "START_TIMER",
      "climateFunction": "PARK_HEATING",
      "climateNow": "ACTIVATED",
      "color": "SCHWARZ  II",
      "colorCode": "668",
      "countryCode": "V1-AT",
      "dealer": {
        "city": "Wien",
        "country": "AT",
        "name": "Wolfgang Denzel Auto AG Standort Erdberg",
        "phone": "+43 1 740080",
        "postalCode": "1030",
        "street": "Erdbergstra\u00dfe 189-193"
      },
      "doorLock": "ACTIVATED",
      "doorUnlock": "ACTIVATED",
      "driveTrain": "CONV",
      "fuelType": "DIESEL",
      "hasAlarmSystem": false,
      "hmiVersion": "ID4",
      "hornBlow": "ACTIVATED",
      "hub": "HUB_ECE",
      "intermodalRouting": "NOT_AVAILABLE",
      "ipa": "NOT_SUPPORTED",
      "lastDestinations": "NOT_SUPPORTED",
      "licensePlate": "some_license_plate",
      "lightFlash": "ACTIVATED",
      "lscType": "NOT_SUPPORTED",
      "model": "520d",
      "onlineSearchMode": "MAP",
      "onlineSearchProvider": "GOOGLE",
      "rangeMap": "NOT_SUPPORTED",
      "remote360": "NOT_SUPPORTED",
      "sendPoi": "ACTIVATED",
      "smartSolution": "NOT_SUPPORTED",
      "statisticsAvailable": false,
      "statisticsCommunityEnabled": false,
      "steering": "LH",
      "vehicleFinder": "ACTIVATED",
      "vehicleFinderRestriction": "NONE",
      "vin": "some_vin",
      "yearOfConstruction": 2015
    }
  ]
}

**Comment by gerard33** [comment]
Yeah that's readable :+1: 

Seems to be the same issue as #62.
Is the data in your BMW ConnectedDrive App also only shown when your phone is within a mile of your car as mentioned [here](https://github.com/m1n3rva/bimmer_connected/issues/62#issuecomment-379439573)?

**Comment by iz3man** [comment]
Well the car is in front of the house where the raspberry is installed :)
I'm on the other side of the planet NOW. But the BMW app works fine in
locating the car. Last update was weeks ago, updated and it shows my car in
front of my house.

Gerard <notifications@github.com> schrieb am Mi., 18. Apr. 2018, 14:12:

**Comment by gerard33** [comment]
> I'm on the other side of the planet NOW.

Definitely another issue then :wink: 

For your car there is indeed some data available, but the code fails on getting the `door_lock_state` which is not available. That is indeed some new code which is added recently.

@m1n3rva I think we should add some checks to the library to avoid  `AttributeError: 'NoneType' object has no attribute 'value'` issues.
Can you have a look at that?

**Comment by iz3man** [comment]
Yes. I guess it tries to gether data which my car doesn't supply and then fails.
Interestingly it doesn't show ANY data if it can't collect all.

**Comment by m1n3rva** [comment]
Hey @gerard33 

having different parameters visible for different vehicle is quite inconvenient :frowning_face: . We somehow have to figure out a way to :
1. identify which vehicle :car: will give which parameters
2. agree on a place to store that information
3. make the home assistant integration flexible :muscle: : it should only show the parameters supported by the vehicle :car: .
4. discuss what the library should do, if you request a parameter that does not exist

So let's get our hands dirty :hammer_and_wrench: :

1. We could have another look at the vehicle attributes. My guess is if `lscType == NOT_SUPPORTED`, you will not get useful information about the vehicle (see example above). At least that's the pattern I've seen so far. But I need to validate :white_check_mark: this with the data we have so far.

2. We did something similar already with the combustion vs. electric parameters. So we could extend this: The library provides a list of parameters the vehicle will (as far as we know) have. Then the Home Assistant components check the list and only create these sensors. Plug-and-play.

3. same as 2.

4. In some cases we return `None` in other we raise an exception. Maybe it's more stable if we only return None if data is not available...

What do you think?

**Comment by gerard33** [comment]
Hi @m1n3rva 

1. I agree with the `lscType`, see my previous [post](https://github.com/m1n3rva/bimmer_connected/issues/62#issuecomment-378409500) on which I suggested that :wink: 
But in #60 there is some data available as you mentioned earlier.
Looks like only the newly added services and device tracker will work in that case however:
```
"climateNow": "ACTIVATED",
"hornBlow": "ACTIVATED",
"lightFlash": "ACTIVATED",

"position": {
      "lat": 12.3456,
      "lon": 34.5678,
      "status": "OK"
    },
```

2. Based on the above I think we only need to check the `lscType`. If it's `NOT_SUPPORTED` then I think we only have to create the device tracker (and the [services](https://github.com/home-assistant/home-assistant/pull/13497), but these are part of the standard component in the next version.

@iz3man, is it possible to share a screenprint of the available data in your BMW ConnectedDrive app with us? I am wondering which data is available in the app for you, is it only device tracker and flashing light, blow horn and turn on the climate?

3. see 2

4. I agree to always return None if data is not available. That will avoid errors and is more user friendly.

**Comment by iz3man** [comment]
![2018-04-24 at 07 12 19](https://user-images.githubusercontent.com/11553674/39241735-362688c4-4888-11e8-9a84-9f1585a8b607.png)
![2018-04-24 at 07 12 15](https://user-images.githubusercontent.com/11553674/39241736-368e1688-4888-11e8-9099-9a80ac2baba9.png)

**Comment by gerard33** [comment]
@iz3man thanks for sharing the screenprints of the app.
I guess despite the message from the first screenshot that you can connect the phone to the car, this will not get you additional info like the mileage?

**Comment by iz3man** [comment]
they changed the app (and it's layout) again.
i connected the phone to my car, and data (mileage, avg fuel consumption etc ..) didn't update on my android. i'm sure this DID work with my iphone. or maybe it takes some time.
btw: mileage and remaining fuel DID show in home assistant. but these were old values. mileage is around 15k lower -  so about 6mo old.

**Comment by iz3man** [comment]
when i connect the phone by usb cable i get information on mileage and remaining fuel which updates just in time.

![2018-04-27 at 14 03 21](https://user-images.githubusercontent.com/11553674/39618873-e96d07c4-4f85-11e8-8e6a-e12656d33c15.png)

**Comment by iz3man** [comment]
btw: i still don't get ANY information at all. HA version 0.68.1.
anything i should try out? i'm not quite sure what the actual status is here. sorry.

```
2018-05-04 07:58:09 ERROR (Thread-18) [bimmer_connected.state] No data available!
2018-05-04 07:58:09 ERROR (Thread-7) [bimmer_connected.state] No data available!
2018-05-04 07:58:09 ERROR (MainThread) [bimmer_connected.state] No data available!
2018-05-04 07:58:09 ERROR (MainThread) [bimmer_connected.state] No data available!
2018-05-04 07:58:09 ERROR (MainThread) [homeassistant.core] Error doing job: Task exception was never retrieved
Traceback (most recent call last):
  File "/usr/lib/python3.5/asyncio/tasks.py", line 239, in _step
    result = coro.send(None)
  File "/srv/homeassistant/lib/python3.5/site-packages/homeassistant/helpers/entity_platform.py", line 316, in _async_add_entity
    await entity.async_update_ha_state()
  File "/srv/homeassistant/lib/python3.5/site-packages/homeassistant/helpers/entity.py", line 223, in async_update_ha_state
    device_attr = self.device_state_attributes
  File "/srv/homeassistant/lib/python3.5/site-packages/homeassistant/components/lock/bmw_connected_drive.py", line 69, in device_sta$
    'door_lock_state': vehicle_state.door_lock_state.value
AttributeError: 'NoneType' object has no attribute 'value'
2018-05-04 07:58:09 ERROR (MainThread) [homeassistant.core] Error doing job: Task exception was never retrieved
Traceback (most recent call last):
  File "/usr/lib/python3.5/asyncio/tasks.py", line 239, in _step
    result = coro.send(None)
  File "/srv/homeassistant/lib/python3.5/site-packages/homeassistant/helpers/entity_platform.py", line 316, in _async_add_entity
    await entity.async_update_ha_state()
  File "/srv/homeassistant/lib/python3.5/site-packages/homeassistant/helpers/entity.py", line 223, in async_update_ha_state
    device_attr = self.device_state_attributes
  File "/srv/homeassistant/lib/python3.5/site-packages/homeassistant/components/binary_sensor/bmw_connected_drive.py", line 94, in d$
    result['door_lock_state'] = vehicle_state.door_lock_state.value
AttributeError: 'NoneType' object has no attribute 'value'
2018-05-04 07:58:09 ERROR (Thread-10) [bimmer_connected.state] No data available!
2018-05-04 07:58:09 ERROR (Thread-21) [bimmer_connected.state] No data available!
2018-05-04 07:58:09 ERROR (Thread-14) [bimmer_connected.state] No data available!
```

**Comment by gerard33** [comment]
> btw: i still don't get ANY information at all. HA version 0.68.1.
anything i should try out? i'm not quite sure what the actual status is here. sorry.

That is not being fixed yet.

@m1n3rva the below error is currently not caught by the `backend_parameter` decorator.
Can you add this `AttributeError` or should we solve this in another way?
```
result['door_lock_state'] = vehicle_state.door_lock_state.value
AttributeError: 'NoneType' object has no attribute 'value'
```

> when i connect the phone by usb cable i get information on mileage and remaining fuel which updates just in time.

Seems that info is only accurately available via the usb connection.
And apparently since we have switched to the other API the mileage and fuel info is not available for you anymore, but I guess that is fine since it well not be helpful to show very out of date info there.

**Comment by m1n3rva** [comment]
Hey @gerard33,

If I read the :bug: report above correctly, the exception is thrown in `homeassistant/components/binary_sensor/bmw_connected_drive.py`, so it's the HA component. I suppose you get `vehicle_state.door_lock_state == None` from `bimmer_connected` as we do not get any data from the server. Calling a `None.value` will raise this exception.

I guess this needs to be changed on the HA side, when we implement the check for the `lscType`. Then you would not create a sensor for which we do not have data from the server...

@iz3man 
I guess that data you see in the :iphone: app is from the last time you connected your phone via USB cable to the vehicle. This data is **not** coming live from the server. So I we will not be able to retrieve any data for your vehicle from the server :cry: But we should be able to fix the error messages in HA :smile:

**Comment by iz3man** [comment]
@m1n3rva you're correct. i mentioned this assumption already some time ago. 
the data i see is from the last time the mobile was connected. i can live with that. the information about mileage is not really helpful. remaining fuel may be useful, but if this ain't work, than that's how it is.
the main issue i see is, that i still have is that there is NO data. so i can't even turn on remote heating/cooling. the senors are there anymore.

**Comment by m1n3rva** [comment]
@gerard33 : I added a first implementation of the `lscTypes` functionality: #74

**Comment by gerard33** [comment]
@iz3man @dws199
Can you check this post https://community.home-assistant.io/t/bmw-connecteddrive-component/39569/262?u=gerard33.
Looks like adding remote services will give you all the data from the car without needing the USB connection.

**Comment by dws199** [comment]
Thanks, I'll check it out

**Comment by m1n3rva** [comment]
closing the ticket as the change was merged #74
